### PR TITLE
Fix letsencrypt

### LINKED ISSUE #60180: nixos/acme: nginx configuration test  fails  during nixos-rebuild switch when you have an existing virtualhost and add a new one
## Issue description

nginx failed to restart after I added a second virtualHost to my nginx config. 
running `nixos-rebuild switch` a second time fixed the issue. But my website went down due to a `nixos-rebuild switch` which seems like there is a serious bug in the `enableACME` logic in the `nginx` module

Friends of me also ran into this multiple times last week at their company

### Steps to reproduce

1. Have a virtualHost with `enableACME = true`
2. Add an extra virtualHost with `enableACME = true` and `nixos-rebuild switch`
3. The `nginx` prestart script (`nginx -t`) fails

This was the specific commit that failed to deploy: https://github.com/arianvp/nixos-stuff/commit/4fe0a777c1874696fd8b985f2b3f68fc01543a55

Logs:

```
Apr 24 14:27:56 arianvp.me systemd[1]: nginx.service: Current command vanished from the unit file, execution of the command list won't be resumed.
Apr 24 14:27:56 arianvp.me systemd[1]: Stopping Nginx Web Server...
-- Subject: Unit nginx.service has begun shutting down
-- Defined-By: systemd
-- Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit nginx.service has begun shutting down.
Apr 24 14:27:56 arianvp.me systemd[1]: Stopped Nginx Web Server.
-- Subject: Unit nginx.service has finished shutting down
-- Defined-By: systemd
-- Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit nginx.service has finished shutting down.
Apr 24 14:27:56 arianvp.me systemd[1]: Starting Nginx Web Server...
-- Subject: Unit nginx.service has begun start-up
-- Defined-By: systemd
-- Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit nginx.service has begun starting up.
Apr 24 14:27:56 arianvp.me s0966gsjr6cilylnniqzh5l35z99jyng-unit-script-nginx-pre-start[18364]: nginx: [emerg] BIO_new_file("/var/lib/acme/techstock.photos/fullchain.pem") fai>
Apr 24 14:27:56 arianvp.me s0966gsjr6cilylnniqzh5l35z99jyng-unit-script-nginx-pre-start[18364]: nginx: configuration file /nix/store/5g561hwc1lw3bbc29qdr83fwl5iykjnv-nginx.con>
Apr 24 14:27:56 arianvp.me systemd[1]: nginx.service: Control process exited, code=exited status=1
Apr 24 14:27:56 arianvp.me systemd[1]: nginx.service: Failed with result 'exit-code'.
Apr 24 14:27:56 arianvp.me systemd[1]: Failed to start Nginx Web Server.
-- Subject: Unit nginx.service has failed
-- Defined-By: systemd
-- Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit nginx.service has failed.
-- 
-- The result is RESULT.
```

## Technical details

Nixos 19.03 but also happened on 18.03 to me in the past

### PR DESCRIPTION
**Pull request by arianvp**
###### Motivation for this change

Fixes https://github.com/NixOS/nixpkgs/issues/60180

Futhermore refactors the module a bit. I'm trying to get rid of `PermissionStartOnly`

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by arianvp** [comment]
I am planning to deprecate the `activationDelay` option as its implementation is a bit confusing, and it can better be replaced with systemd dependencies instead of waiting for an arbitrary time. (e.g. what happens if the time passes, but whatever script you run during the delay didn't finish??).


I think a good replacement instead would be to create a  `acme-wait-dns-published.service`   yourself and then do:

```
# acme-wait-dns-published.service
RequiredBy=acme-${cert}.service
After=acme-${cert}.service
```

This way, the delay isn't an arbitrary number, but mandated by the script that you use to publish to `DANE/TLSA`. 
and it greatly simplifies the letsencrypt activation logic, which is now hard to follow.

See https://github.com/NixOS/nixpkgs/commit/e70d293b6b5f236d69d559ccccdafba19c6d29c3#commitcomment-34131909 for context

**Comment by arianvp** [comment]
CC @pngwjpgh

**Comment by grahamc** [comment]
SGTM. Any fixed "delay" value is inherently wrong.

**Comment by arianvp** [comment]
One big problem:

You can't `waitForUnit` on `oneshot` services, as their state goes from `inactive => activating => inactive`. This makes the tests currently fail.

If anybody has an idea on how to test for oneshots firing, please let me know

**Comment by grahamc** [comment]
Even with remain after exit?

**Comment by arianvp** [comment]
RemainAfterExit will work but then the unit remains active which means the
timer unit will not periodically activate it causing your certs to expire :p

**Comment by arianvp** [comment]
@grahamc I've fixed the test by letting each oneshot activate a target (Which remains active after running) to use as a probe to test for unit activation. It's a bit hacky, but it works!

File: nixos/modules/rename.nix [hunk 1]
```diff
@@ -240,6 +240,11 @@ with lib;
 
     # binfmt
     (mkRenamedOptionModule [ "boot" "binfmtMiscRegistrations" ] [ "boot" "binfmt" "registrations" ])
+    
+    # ACME
+    (mkRemovedOptionModule [ "security" "acme" "directory"] "ACME Directory is now hardcoded to /var/lib/acme and its permisisons are managed by systemd. See https://github.com/NixOS/nixpkgs/issues/53852 for more info.")
```
Comment (Mic92) [inline nixos/modules/rename.nix]:
```suggestion
    (mkRemovedOptionModule [ "security" "acme" "directory"] "ACME Directory is now hard-coded to /var/lib/acme and its permissions are managed by systemd. See https://github.com/NixOS/nixpkgs/issues/53852 for more info.")
```

Comment (arianvp) [reply nixos/modules/rename.nix]: I was wondering whether this line even works in the first place.
I did not remove the `directory` option. I made it `readOnly` instead.

I honestly have no clue how the `rename.nix` module works. so perhaps this is totally wrong in the first place?

**Comment by arianvp** [comment]
@GrahamcOfBorg test acme

**Comment by arianvp** [comment]
A few problems remain:

1. certificates add dependencies to `nginx` and `lighttpd` eventhough they're not used by said services
2.  `lighttpd` seems to have no direct ACME enable support whatsoever so it seems weird to make `lighttpd.service` pull in all these certs even if it doesn't use them

2. was introduced by @bjornfor in https://github.com/NixOS/nixpkgs/commit/7a0e958b972f61b23f93a8e39fee5590aec47348 

Would you mind if I remove this coupling altogether, and document instead that people
can do:
```
mywebserver.wants  =  [ "acme-${cert}.service acme-selfsigned-${cert}.service" ];
mywebserver.after = ["acme-selfsigned-${cert}.service"];
```
whenever you want to depend on certs being generated on startup?


For `nginx` we can probably do this automatically when a `vhost` has `enableACME  = true;` but lets make sure it only pulls in the certs for that specific vhost instead of _all_ certs

This will also fix https://github.com/NixOS/nixpkgs/issues/62958

**Comment by bjornfor** [comment]
@arianvp: I don't mind. I only added that line because nginx had it, and it felt like the natural thing to do at that time.

**Comment by arianvp** [comment]
This is ready for review.  I can refactor even more but lets keep that for a new PR or a new NixOS Module altogether. At least the bugs that were present before are fixed now.

**Comment by arianvp** [comment]
I think we should do the big refactor to DNS challenges and `lego` in a separate PR and maybe even in a separate NixOS module. it's not worth keeping backwards compatibility with what we have currently I think. It might hamper us coming up with better designs. @mweinelt would you perhaps want to take a look at this PR?

File: nixos/modules/security/acme.nix [hunk 2]
```diff
@@ -221,82 +199,32 @@ in
                   description = "Renew ACME Certificate for ${cert}";
                   after = [ "network.target" "network-online.target" ];
                   wants = [ "network-online.target" ];
+                  wantedBy =
+                    lib.optionals config.services.nginx.enable  [ "nginx.service" ] ++
```
Comment (nh2) [inline nixos/modules/security/acme.nix]: Does it make sense that his is here?

Given that the `enableACME` option is in the nginx module, shouldn't that one use `wants` on the cert services instead?

It looks like this way, any webserver present in nixpkgs that might use these certs would have to add itself here.

If doing it the other way around is currently not feasible (e.g. because it would be somehow more difficult or inconvenient), can you add a comment stating so?

Comment (arianvp) [reply nixos/modules/security/acme.nix]: Also see discussion above. I was planning to do this but forgot. Thanks for reminding me

Comment (arianvp) [reply nixos/modules/security/acme.nix]: I removed these now!

**Review by flokli — APPROVED** [review]

File: nixos/modules/rename.nix [hunk 3]
```diff
@@ -240,6 +240,11 @@ with lib;
...
+    (mkRemovedOptionModule [ "security" "acme" "preDelay"] "TODO(arianvp): Write removed notice")
+    (mkRemovedOptionModule [ "security" "acme" "activationDelay"] "TODO(arianvp): Write removed notice")
```
Comment (Mic92) [inline nixos/modules/rename.nix]: TODO should be fixed before merging

File: nixos/modules/rename.nix [hunk 4]
```diff
@@ -240,6 +240,11 @@ with lib;
 
     # binfmt
     (mkRenamedOptionModule [ "boot" "binfmtMiscRegistrations" ] [ "boot" "binfmt" "registrations" ])
+    
+    # ACME
+    (mkRemovedOptionModule [ "security" "acme" "directory"] "ACME Directory is now hardcoded to /var/lib/acme and its permisisons are managed by systemd. See https://github.com/NixOS/nixpkgs/issues/53852 for more info.")
+    (mkRemovedOptionModule [ "security" "acme" "preDelay"] "TODO(arianvp): Write removed notice")
```
Comment (Mic92) [inline nixos/modules/rename.nix]: TODO should be fixed before merging

Comment (arianvp) [reply nixos/modules/rename.nix]: I got rid of the TODOs

**Comment by flokli** [comment]
Ok, there was some wrong escaping of `${cert}`, and a missing `</para>`, which caused the manual build to fail. Should be good now :-)

**Comment by flokli** [comment]
Due to the module system changes, this might need an entry in `nixos/doc/manual/release-notes/rl-1909.xml`.

**Comment by aanderse** [comment]
This needs to be reviewed again. I haven't done any testing, or even checked that the manual builds (as the manual is currently broken in `master` I believe). I'm also not familiar with `nginx`, so someone should scrutinize the (minor) changes I've made, just to be sure.

@arianvp would be nice to get your :+1: on my changes, as it is your PR :smile:

**A squash would also be nice at this point.**

**Comment by arianvp** [comment]
@GrahamcOfBorg test acme

**Comment by arianvp** [comment]
LGTM; thanks for lending me a hand.

File: nixos/doc/manual/release-notes/rl-1909.xml [hunk 5]
```diff
@@ -284,6 +284,16 @@
        Squid 3 has been removed and the <option>squid</option> derivation now refers to Squid 4.
      </para>
    </listitem>
+   <listitem>
+     <para>
+       The <option>security.acme.directory</option> option has been replaced by a read-only <literal>directory</literal> option for each certificate you define. The
+       <option>security.acme.preDelay</option> and <option>security.acme.activationDelay</option> options have been removed. To execute a service before certificates
+       are provisioned or renewed add a <literal>RequiredBy=acme-${cert}.service</literal> to any service.
```
Comment (flokli) [inline nixos/doc/manual/release-notes/rl-1909.xml]: We should add that this is already done if `services.nginx.virtualHosts.<name>.enableACME` is set to `true`, so people don't wonder if they now need to add this to their nginx configuration.

**Review by flokli — CHANGES_REQUESTED** [review]
Can you rebase this on latest master?

File: nixos/doc/manual/release-notes/rl-1909.xml (hunk 5 above)
Comment (arianvp) [reply nixos/doc/manual/release-notes/rl-1909.xml]: rebased and added to the docs

File: nixos/modules/security/acme.nix [hunk 6]
```diff
@@ -181,7 +160,11 @@ in
         default = { };
         type = with types; attrsOf (submodule certOpts);
         description = ''
-          Attribute set of certificates to get signed and renewed.
+          Attribute set of certificates to get signed and renewed. Creates
+          <literal>acme-''${cert}.{service,timer}</literal> systemd units for
+          each certificate defined here. Other services can then add
+          dependencies on the created service file to do something when a
```
Comment (flokli) [inline nixos/modules/security/acme.nix]: what should they do?

What about
> can add dependencies to those units if they rely on the certificates being present, or trigger restarts of the service if certificates get renewed.

File: nixos/doc/manual/release-notes/rl-1909.xml [hunk 7]
```diff
@@ -318,7 +318,28 @@
...
+     <para>
+       <option>security.acme.preDelay</option> and <option>security.acme.activationDelay</option> options have been removed. To execute a service before certificates
+       are provisioned or renewed add a <literal>RequiredBy=acme-${cert}.service</literal> to any service.
+     </para>
+     <para>
+       Furthermore, the acme module will not automatically add a dependency on <literal>lighttpd.service</literal> anymore. If you are using certficates provided by letsencrypt
+       for lighttpd, then you should depend on the certificate service <literal>acme-${cert}.service></literal> manually.
+     </para>
+     <para>
+       For nginx, the dependencies are still automatically managed when <option>services.nginx.virtualhosts.&lt;name&gt;.enableACME</option>       is enabled just like before. What changed is that nginx now directly depends on the specific certificates that it needs,
```
Comment (flokli) [inline nixos/doc/manual/release-notes/rl-1909.xml]:
```suggestion
       For nginx, the dependencies are still automatically managed when <option>services.nginx.virtualhosts.&lt;name&gt;.enableACME</option> is enabled just like before. What changed is that nginx now directly depends on the specific certificates that it needs,
```

many whitespaces

File: nixos/doc/manual/release-notes/rl-1909.xml [hunk 8]
```diff
@@ -318,7 +318,28 @@
...
+       instead of depending on the catch-all <literal>acme-certificates.target</literal>.  This target unit was also removed from the codebase.
```
Comment (flokli) [inline nixos/doc/manual/release-notes/rl-1909.xml]:
```suggestion
       instead of depending on the catch-all <literal>acme-certificates.target</literal>. This target unit was also removed from the codebase.
```

whitespace

File: nixos/doc/manual/release-notes/rl-1909.xml [hunk 9]
```diff
@@ -318,7 +318,28 @@
...
+       This will mean nginx will no longer depend on certificates it isn't explicitly managing and fixes a bug with certificate renewal
+       ordering racing with nginx restarting which could lead to nginx getting in a broken state as described at 
+        <link xlink:href="https://github.com/NixOS/nixpkgs/issues/60180">NixOS/nixpkgs#60180</link>.
+     </para>
```
Comment (flokli) [inline nixos/doc/manual/release-notes/rl-1909.xml]: :heart:

**Comment by flokli** [comment]
Thanks!

**Comment by Shados** [comment]
FWIW, I was actually using `services.acme.directory`, so removing that has given me some grief >.>.

**Comment by arianvp** [comment]
Feel free to come up with an alternative. We could use `systemd.tmpfiles` for example.   however don't personally have time to work on this.  But I'm happy to review the PR that reintroduces that option

**Comment by Shados** [comment]
Not sure `tmpfiles` is the right approach either... the usage of it in the refactored module is also somehow not working right for me -- the activation script snippet for it is failing with "Detected unsafe path transition" =/. I'll take a look into this tonight.

**Comment by arianvp** [comment]
This is error is generated because of:

```
        /* Returns true if the transition from a to b is safe, i.e. that we never transition from unprivileged to
         * privileged files or directories. Why bother? So that unprivileged code can't symlink to privileged files
         * making us believe we read something safe even though it isn't safe in the specific context we open it in. */
```
https://github.com/systemd/systemd/blob/6fd79cca68a170c88e87b5d95cd7a44953df94e2/src/basic/fs-util.c#L663

Does that help with debugging the issue?

**Comment by aanderse** [comment]
1. Entirely out of curiosity: what directory are you using, and will symlinks not solve your problem?
2. If `tmpfiles.d` is not explicitly told who ownsa directory it will assign `root` as the owner. This includes parent directories of the directory you specify. Under certain permission schemes `tmpfiles.d` decides this is not safe (the right choice) and throws the error/warning. See https://github.com/NixOS/nixpkgs/pull/67888 for concrete example.

I'm happy to hear that what and why of your custom directory to change my opinion, but as it currently stands if symlinks would do the job I continue to agree with the decision to remove the directory option.

**Comment by Shados** [comment]
@arianvp Hmm, yeah, I why my setup is causing the warning now. My config predates `services.nginx.virtualHosts.<name>.enableACME` existing, so the link between the acme webroot and the nginx one was manually defined, and the ownership of that portion of the directory tree ended up being a little odd. I'll change it to use the newer stuff, which is nicer, so that's good.

@aanderse Sure, I can detail the reasoning a bit: essentially, I keep any service-related data that is worth backing up* under a directory tree rooted at `/srv`, partly to simplify administration, partly 

*: By "worth backing up", I mean "not trivially recreat

**Comment by Shados** [comment]
@arianvp Hmm, yeah, I why my setup is causing the warning now. My config predates `services.nginx.virtualHosts.<name>.enableACME` existing, so the link between the acme webroot and the nginx one was manually defined, and the ownership of that portion of the directory tree ended up being a little odd. I'll change it to use the newer stuff, which is nicer, so that's good.

@aanderse Sure, I can detail the reasoning a bit: essentially, I keep any service-related data that is worth backing up* under a directory tree rooted at `/srv`, which simplifies administration in a few ways (e.g. I don't have to back up different service directories per-server, I can configure filesystem/disk-related stuff for all relevant service data on a system in a single place).

`/var/lib` doesn't qualify for this by because it is just generically for service/application state, not specifically for stuff I consider valuable, meaning it inevitably ends up containing a bunch of random stuff I don't care about and don't want consuming backup space. The semantics don't match up.

Could I use a symlink? Often. But then I have more problems:
- If I decide to manually create the symlink, I now have to remember to do that if I am ever moving or replicating this configuration elsewhere.
- If I decide to write a service or activation script or something else to ensure the symlink exists, I then have to test that, ensure it has the appropriate dependencies, and try to cover any edge cases that might come up.
- I can always symlink from `/srv/<something>` to elsewhere, but I can't always do the reverse -- the application in question may not like being given a symlink in its path, for one reason or another, so instead I'd have to use a bind mount (which then has its own potential issues).

Pointing the service in question directly to the directory I actually want it to use is always going to be the better option.


*: By "worth backing up", I mean "not trivially reproduceable". LE certs fall under this classification because acquiring them is rate-limited, so having to recreate them en masse would potentially be problematic. Combine that with HTTP Strict Transport Security and you could end up with effectively inaccessible domains for a while.

**Comment by aanderse** [comment]
@Shados to address your first two points: if you use `systemd.tmpfiles.rules` to create the symlink and those problems disappear. I see your third point as something worth serious consideration, but on a service by service basis.

I'm not against having an option for specifying the state directory when there is a justification for such an option (like an application that can't deal with symlinks as you mentioned), but it would be nice if there **was** justification, as opposed to preference. But then again... this is just **my** opinion, and I don't make the rules.

**Comment by Shados** [comment]
@aanderse I got around to taking the symlink approach with `systemd.tmpfiles.rules`. It works, but it is deceptively complicated, and it quickly became apparent that this is really not what `systemd-tmpfiles` is designed for. Here's a bit of a post-mortem, so to speak:

Initially, I went with just a directory entry for `/srv/acme` (with the same user and owner that I have the acme services running as), and a symlink entry for `/var/lib/acme -> /srv/acme`, but this caused an issue.

Namely, I was back to getting the "unsafe path transition" warnings. And while this is just a *warning* for `systemd-tmpfiles`, and it still creates the path(s), it is an *error* for NixOS because the non-0 return code causes the activation script snippet to fail, which means the system profile activation does not continue.

The cause of the warning was: `systemd-tmpfiles` will ensure that `leading directories are implicitly created if needed`, but they will be `owned by root with an access mode of 0755`. As a result, there was an unsafe path transition going from `/srv/acme` to `/srv/acme/acme-challenge`, because the latter is `implicitly created` as root on the way to making `/var/lib/acme/acme-challenge/.well-known/acme-challenge` (through the symlink).

`man tmpfiles.d` says `In order to create [the leading directories] with different modes or ownership make sure to add appropriate d lines`, so I wrote a Nix helper to map a portion of a path to a list of largely identical `systemd.tmpfiles.rules` directory creation lines. I do note an issue with this approach currently:
If you have two separate entries creating the same directory with different permissions or ownership, you'll get a warning on stderr to that effect during activation, but it appears the return code is still 0, so it is non-obvious.
We could solve that in NixOS by making `systemd.tmpfiles.rules` a structured `attrsOf tmpfileRuleType` instead of a `listOf str`, and I would strongly suggest this be done if the recommendation going forward is to handle service/etc. directory creation using `systemd.tmpfiles.rules`. It would also then be possible to catch some potential "unsafe path transitions" at eval-time.

All up: I do think that using a tool designed to declaratively manipulate a filesystem is the right approach to solving this kind of issue in NixOS, but that using `systemd-tmpfiles` as-is may not be. It ended up taking about as much testing and handling of edge cases as I expected a hand-written, imperative approach would have.
